File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    seen = Set.new()
    left = result = 0

    s.chars.each_index do |right|
        while seen.include?(s[right])
            seen.delete(s[left])
            left += 1
        end
                        
        seen.add(s[right])
        result = [result, right - left + 1].max
    end

    result
end

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        cset = set()
        left = result = 0

        for right in range(len(s)):
            while s[right] in cset:
                cset.remove(s[left])
                left += 1
            cset.add(s[right])
            result = max(result, right - left + 1)
        return result